plugins/logitech-bulkcontroller/bulk_controller.c

```diff
@@ -0,0 +1,879 @@
+#include "bulk_util.h"
+
+void*                   logibulkcontroller_thread_write_buffer          (void* thread_data);
+void*                   logibulkcontroller_thread_send_data             (void *thread_data);
+void*                   logibulkcontroller_thread_read_data             (void *thread_data);
+void*                   logibulkcontroller_thread_send_upd_file         (void* thread_data);
+void*                   logibulkcontroller_thread_read_sync_data        (void* data);
+static gboolean         logibulkcontroller_is_sync_idle                 (void);
+static gint             logibulkcontroller_init_transfer                (void);
+static gint             logibulkcontroller_start_transfer               (GBytes* file_data);
+static gint             logibulkcontroller_finish_transfer              (void);
+
+typedef struct _LogiBulkControllerPrivate
+{
+        BulkControllerDevice                    ctrl_device;
+
+        gint                                    packet_length;
+        guchar                                  base64hash[BASE64_LENGTH];
+        gboolean                                do_update;
+        gboolean                                is_transfer_in_progress;
+        gboolean                                is_upd_transfer_in_progress;
+        guint32                                 transaction_id;
+        guint                                   file_transfer_status;
+        gint                                    file_size;
+        gchar                                   *data_received_async;
+
+        UsbPacket                               data_packet;
+        UsbPacketSync                           data_packet_sync;
+        UsbPacket                               recv_pckt;
+
+        GThread                                 *send_file_upd_thread;
+        GThread                                 *read_sync_data_thread;
+        GThread                                 *read_buffer_thread;
+        GThread                                 *send_data_thread;
+        GThread                                 *write_buffer_thread;
+
+        GQueue                                  *write_buffer_queue;
+        GQueue                                  *read_data_queue;
+        GQueue                                  *send_data_queue;
+
+        Lock                                    lock[THREAD_COUNT];
+        GBytes                                  *file_buffer;
+        ReturnValue                             ret;
+        ApiUserData                             *user_data;
+        BulkControllerCallbacks                 callback;
+}LogiBulkControllerPrivate;
+
+LogiBulkControllerPrivate* priv;
+static void logibulkcontroller_start_listening_sync (void);
+static void logibulkcontroller_stop_listening_sync  (void);
+static void logibulkcontroller_finalize             (GObject *obj);
+
+G_DEFINE_TYPE_WITH_CODE (LogiBulkController, logibulkcontroller, G_TYPE_OBJECT, G_ADD_PRIVATE(LogiBulkController));
```

Comment:
G_DEFINE_TYPE_WITH_PRIVATE if the object is derivable, or just G_DEFINE_TYPE if it's final.

Reply:
Done